src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/RazorHtmlWriter.cs

```diff
@@ -14,311 +14,394 @@ namespace Microsoft.AspNetCore.Razor.Language;
     {
-        WriteNode(node, isHtml: true, _baseVisitMarkupTagHelperStartTag);
+        using (HtmlScope())
+        {
+            base.VisitMarkupTagHelperStartTag(node);
+        }
     }
 
     public override void VisitMarkupTagHelperEndTag(MarkupTagHelperEndTagSyntax node)
     {
-        WriteNode(node, isHtml: true, _baseVisitMarkupTagHelperEndTag);
+        using (HtmlScope())
+        {
+            base.VisitMarkupTagHelperEndTag(node);
+        }
     }
 
     public override void VisitMarkupEphemeralTextLiteral(MarkupEphemeralTextLiteralSyntax node)
     {
-        WriteNode(node, isHtml: true, _baseVisitMarkupEphemeralTextLiteral);
+        using (HtmlScope())
+        {
+            base.VisitMarkupEphemeralTextLiteral(node);
+        }
     }
 
     public override void VisitMarkupTextLiteral(MarkupTextLiteralSyntax node)
     {
-        WriteNode(node, isHtml: true, _baseVisitMarkupTextLiteral);
+        using (HtmlScope())
+        {
+            base.VisitMarkupTextLiteral(node);
+        }
     }
 
     public override void VisitUnclassifiedTextLiteral(UnclassifiedTextLiteralSyntax node)
     {
-        WriteNode(node, isHtml: true, _baseVisitUnclassifiedTextLiteral);
+        using (HtmlScope())
+        {
+            base.VisitUnclassifiedTextLiteral(node);
+        }
     }
 
     public override void VisitToken(SyntaxToken token)
     {
-        base.VisitToken(token);
-        WriteToken(token);
+        if (_isWritingHtml)
+        {
+            WriteHtmlToken(token);
+        }
+        else
+        {
+            WriteNonHtmlToken(token);
+        }
     }
 
-    private void WriteToken(SyntaxToken token)
+    private readonly ref struct WriterScope
     {
-        var content = token.Content;
-        if (_isHtml)
+        private readonly RazorHtmlWriter _writer;
+        private readonly bool _oldIsWritingHtml;
+
+        public WriterScope(RazorHtmlWriter writer, bool isWritingHtml)
         {
-            WriteDeferredCSharpContent();
+            _writer = writer;
+            _oldIsWritingHtml = writer._isWritingHtml;
+            writer._isWritingHtml = isWritingHtml;
+        }
 
-            var source = token.GetSourceSpan(Source);
+        public void Dispose()
+        {
+            _writer._isWritingHtml = _oldIsWritingHtml;
+        }
+    }
 
-            // No point source mapping an empty token
-            if (source.Length > 0)
-            {
-                var generatedLocation = new SourceSpan(Builder.Location, source.Length);
+    private WriterScope NonHtmlScope()
+        => new(this, isWritingHtml: false);
 
-                if (_lastGeneratedSourceSpan == SourceSpan.Undefined)
-                {
-                    // Not tracking any current source mapping, so start tracking one
+    private WriterScope HtmlScope()
+        => new(this, isWritingHtml: true);
+    private void WriteHtmlToken(SyntaxToken token)
```

Comment:
nit: consider adding a blank line between the declarations

```suggestion
    private WriterScope HtmlScope()
        => new(this, isWritingHtml: true);

    private void WriteHtmlToken(SyntaxToken token)
```

Reply:
Derp! thanks!